File: vipaq/packages/binacle-vipaq/src/compression/deflateCodec.ts
import {CompressionCodec} from "./compressionCodec";
import {runStream} from "./runStream";
import {ViPaqFormatError} from "../utils";

// Ports C#: DeflateCodec. Raw DEFLATE (RFC 1951), the same stream C#'s DeflateStream writes. The '-raw' variant
// is required: plain 'deflate' adds a zlib header (RFC 1950) that C# does not produce, so it would not interop.
// Node exposes CompressionStream as a global, so this one path serves the browser and Node.
//
// Compressed bytes are NOT compared across engines - different zlib builds differ. The cross-language guarantee
// is decode-to-input (PROTOCOL.md §6.1), proven by the round-trip tests, not taken on trust.
export const deflateCodec: CompressionCodec = {
	async compress(body: Uint8Array): Promise<Uint8Array<ArrayBuffer>> {
		return runStream(body, new CompressionStream("deflate-raw"));
	},

	async decompress(compressed: Uint8Array): Promise<Uint8Array<ArrayBuffer>> {
		try {
			return await runStream(compressed, new DecompressionStream("deflate-raw"));
		} catch (error) {
			throw new ViPaqFormatError("The compressed body is not a valid DEFLATE stream");
		}
	},
};

File: vipaq/src/Binacle.ViPaq/Compression/DeflateCodec.cs
using System.IO.Compression;

namespace Binacle.ViPaq.Compression;

// Raw DEFLATE (RFC 1951), no wrapper. Cheaper than gzip by ~18 bytes a blob, which pushes the point where
// compressing starts to pay closer in.
//
// `DeflateStream` pairs with `CompressionStream('deflate-raw')` in a browser and `zlib.createDeflateRaw` in
// Node. That is only proven by the cross-language round-trip tests - do not take it on trust.
internal sealed class DeflateCodec : ICompressionCodec
{
	public byte[] Compress(ReadOnlySpan<byte> body)
	{
		using var output = new MemoryStream();
		using (var deflate = new DeflateStream(output, CompressionLevel.Optimal, leaveOpen: true))
		{
			deflate.Write(body);
		}

		return output.ToArray();
	}

	public byte[] Decompress(ReadOnlySpan<byte> compressed)
	{
		using var input = new MemoryStream(compressed.ToArray());
		using var deflate = new DeflateStream(input, CompressionMode.Decompress);
		using var output = new MemoryStream();

		try
		{
			deflate.CopyTo(output);
		}
		catch (InvalidDataException exception)
		{
			throw new ViPaqFormatException("The compressed body is not a valid DEFLATE stream", exception);
		}

		return output.ToArray();
	}
}
